// neo/tabs/story/index.js
const path = require('path');

const log = require('log');
const neo = require('neo');

const $ = require('jquery');

const {clipboard} = require('electron');

self = module.exports = {
  get StoryFolders() {
    return [path.join(__dirname, 'docs')];
  },

  get name() {
    return 'story';
  },

  get editor() {
    return $(neo.StoryPane).find('story-editor')[0];
  },

  openTab() {
    var workDiv = $(neo.StoryPane).find('nui-work');
    var editor = document.createElement('story-editor');

    editor.story = neo.CurrentStory;
    $(workDiv).html(editor);

    // !!!TODO!!! not pleased with this method of setting focus, but it was the least-bad alternative.
    setTimeout(() => $(workDiv).find('story-editor')[0].focus(), 0);
  },

  closeTab() {
    var workDiv = $(neo.StoryPane).find('nui-work');

    $(workDiv).find('story-editor').remove();
  },

  Commands: [
    {
      name: 'cursor left',
      fn: function () {
        this.editor.doc.cursorLeft();
      }
    },
    {
      name: 'cursor right',
      fn: function () {
        this.editor.doc.cursorRight();
      }
    },
    {
      name: 'cursor up',
      fn: function () {
        this.editor.doc.cursorUp();
      }
    },
    {
      name: 'cursor down',
      fn: function () {
        this.editor.doc.cursorDown();
      }
    },
    {
      name: 'drop anchor',
      fn: function () {
        this.editor.doc.dropAnchor();
      }
    },
    {
      name: 'weigh anchor',
      fn: function () {
        this.editor.doc.weighAnchor();
      }
    },
    {
      name: 'select all',
      fn: function () {
        this.editor.doc.selectAll();
      }
    },
    {
      name: 'copy',
      fn: function () {
        var range = this.editor.doc.selectionRange;
        if (range) {
          var div = document.createElement('div');
          div.appendChild(range.cloneContents());

          clipboard.clear();
          clipboard.writeHTML(div.innerHTML);
          clipboard.writeText(div.textContent);
        }
      }
    },
    {
      name: 'save',
      fn: function () { this.editor.save(); }
    }
  ]
}

function initTab() {
  var moduleLink = $('<link rel="import" href="' + path.join(__dirname, 'index.html') + '"></link>');
  $('head').append(moduleLink);

  moduleLink = $('<link rel="import" href="' + path.join(__dirname, 'neo-story.html') + '"></link>');
  $('head').append(moduleLink);
};

initTab();
