// neo/tabs/story/index.js
const path = require('path');

const log = require('log');
const neo = require('neo');

const $ = require('jquery');

self = module.exports = {
  get StoryFolders() {
    return [path.join(__dirname, 'docs')];
  },

  get name() {
    return 'story';
  },

  get editor() {
    return $(neo.StoryPane).find('story-editor')[0];
  },

  openTab() {
    var workDiv = $(neo.StoryPane).find('nui-work');
    var editor = document.createElement('story-editor');

    editor.story = neo.CurrentStory;
    $(workDiv).html(editor);

    // !!!TODO!!! not pleased with this method of setting focus, but it was the least-bad alternative.
    setTimeout(() => $(workDiv).find('story-editor')[0].focus(), 0);
  },

  closeTab() {
    var workDiv = $(neo.StoryPane).find('nui-work');

    $(workDiv).find('story-editor').remove();
  },

  Commands: [
    {
      name: 'left',
      fn: function () {
        document.querySelector('story-editor').doc.cursorLeft();
      }
    },
    {
      name: 'right',
      fn: function () {
        document.querySelector('story-editor').doc.cursorRight();
      }
    }
  ]
}

function initTab() {
  var moduleLink = $('<link rel="import" href="' + path.join(__dirname, 'index.html') + '"></link>');
  $('head').append(moduleLink);

  moduleLink = $('<link rel="import" href="' + path.join(__dirname, 'neo-story.html') + '"></link>');
  $('head').append(moduleLink);
};

initTab();
