// neo/tabs/about/index.js
// !!!LATER!!! this entire about framework is a placeholder, once I have better
// requirements for metadata and other-tab participation it will be refactored.
const path = require('path');

const log = require('log');
const neo = require('neo');

const $ = require('jquery');

self = module.exports = {
  get StoryFolders() {
    return [];
  },

  get name() {
    return 'about';
  },

  openTab() {
    var story = neo.CurrentStory;

    var workDiv = $(neo.StoryPane).find('nui-work');
    var editor = $($('#about_editor')[0].import).find('.about_editor')[0].outerHTML;

    $(workDiv).html(editor);
    $(workDiv).find('.about_editor').on('keyup', updateMetadata);
    $(workDiv).find('.about_editor').on('change', updateMetadata);

    $(workDiv).find('.about_editor h1').text(story.metadata.title);

    var form = $(workDiv).find('.about_editor form')[0];
    form.title.value = story.metadata.title;
    form.keywords.value = story.metadata.keywords.join(', ');
  },

  closeTab() {
    var workDiv = $(neo.StoryPane).find('nui-work');

    $(workDiv).find('.about_editor').remove();
  }
}

function updateMetadata() {
  var story = neo.CurrentStory;
  var workDiv = $(neo.StoryPane).find('nui-work');

  var form = $(workDiv).find('.about_editor form')[0];

  story.metadata.title = form.title.value;

  var keywords = form.keywords.value;

  story.metadata.keywords = keywords.split(',').map(kw => kw.trim());

  $(workDiv).find('.about_editor h1').text(story.metadata.title);

  story.write();
}

function initTab() {
  var moduleLink = $('<link id="about_editor" rel="import" href="' + path.join(__dirname, 'index.html') + '"></link>');
  $('head').append(moduleLink);
};

initTab();
