// neo/Author.js
const JSONPackageFile = require('./JSONPackageFile.js')

const typeURL = 'https://github.com/smeans/neo#NeoAuthorFile';

module.exports = class Author {
  /** Check if given path is a neo author profile. */
  static isAuthor(filePath) {
    // !!!LATER!!! add verification of type URL
    return ['.naf'].indexOf(path.extname(filePath).toLowerCase()) > -1;
  }

  /** Construct a new Author from a JSONPackageFile path. */
  constructor(filePath) {
    this.jpf = new JSONPackageFile(filePath, typeURL);
  }

  /** Attempts to return the author's first name. */
  get firstName() {
    var names = this.metadata.fullName.split(/\s+/);
    return names[0];
  }

  get metadata() {
    return this.jpf.getChunk('jpf_metadata');
  }

  getPref(key) {
    var prefs = this.jpf.getChunk('neo_prefs');

    return prefs.values && prefs.values[key];
  }

  setPref(key, val) {
    var prefs = this.jpf.getChunk('neo_prefs');
    if (!('values' in prefs)) {
      prefs.values = {};
    }

    prefs.values[key] = val;

    this.jpf.write();
  }

  updateChunk(chunkType, fn) {
    this.jpf.updateChunk(chunkType, fn);
  }
}
