// neo/neo.js
'use strict';

const path = require('path');
const fs = require('fs-extra');

const electron = require('electron');

const log = require('log');

const User = require('./User.js');
const Author = require('./Author.js');
const Story = require('./Story.js');

/**
 *  Main shared module for neo word processor.
 *
 *  @module neo
 */

var _neoRoot;
var _CurrentUser = null;
var _CurrentAuthor = null;
var _CurrentStory  = null;
var _StoryPane = null;
var _Tabs = {};

var self = module.exports = {
  /** Class representing the neo user. There can be only one user, but a single
  user can have many Author profiles. */
  User,

  /** Class representing a neo author. */
  Author,

  /** Class representing a neo story. */
  Story,

  /** Current neo root folder. This is the container for all User, Author, and
  Story data. */
  get neoRoot() {
    return _neoRoot;
  },

  set neoRoot(newRoot) {
    _neoRoot = newRoot;
    log.debug('new neo root: ' + newRoot);
    loadRoot(_neoRoot);
  },

  /** Current active user profile. */
  get CurrentUser() {
    return _CurrentUser;
  },

  set CurrentUser(newUser) {
    _CurrentUser = newUser;

    fs.writeJsonSync(path.join(_neoRoot, '/user.nuf'), _CurrentUser);
  },

  /** Current active author profile. */
  get CurrentAuthor() {
    return _CurrentAuthor || _CurrentUser.DefaultAuthor;
  },

  set CurrentAuthor(newAuthor) {
    _CurrentAuthor = newAuthor;
  },

  /** Current active story. */
  get CurrentStory() {
    return _CurrentStory;
  },

  set CurrentStory(newStory) {
    _CurrentStory = newStory;
  },

  /** Pane where current story will be displayed/edited. */
  get StoryPane() {
    return _StoryPane;
  },

  set StoryPane(newPane) {
    _StoryPane = newPane;
  },

  /** Start a new story. */
  newStory
}

function loadRoot(root) {
  fs.ensureDirSync(root);
  try {
    _CurrentUser = new User(fs.readJsonSync(path.join(root, '/user.nuf')));

    // load built-in tabs
    loadTabs(path.join(__dirname, 'tabs'));

    // load install-specific tabs
    loadTabs(path.join(root, 'tabs'));
  } catch (e) {}
}

function loadTabs(tabRoot) {
  _Tabs = {};

  fs.readdirSync(tabRoot).forEach(tabName => {
    log.debug('tab: ' + tabName);
    try {
      if (!(tabName in _Tabs)) {
        _Tabs[tabName] = require(path.join(tabRoot, tabName));
      } else {
        log.warn('not loading tab ' + tabName + ': conflict with built-in tab');
      }
    } catch (e) {
      log.error('unable to load tab ' + tabName, e);
    }
  });
}

function newStory() {
  self.CurrentStory = new Story(self.CurrentAuthor);
}
