// neo/tabs/darlings/index.js
const path = require('path');

const log = require('log');
const neo = require('neo');

const $ = require('jquery');

self = module.exports = {
  get StoryFolders() {
    return [];
  },

  get name() {
    return 'darlings';
  },

  openTab() {
    var story = neo.CurrentStory;

    var workDiv = $(neo.StoryPane).find('nui-work');
    var editor = $($('#darlings_view')[0].import).find('.darlings_view')[0].outerHTML;

    var storyBody = neo.CurrentStory.story.body;
    var darlings = $(storyBody)[0].querySelectorAll('nui-darling');

    $(workDiv).html($(editor));
    $(workDiv).find('.darlings_list').append(darlings);
  },

  closeTab() {
    var workDiv = $(neo.StoryPane).find('nui-work');

    $(workDiv).find('.darlings_view').remove();
  },

  Commands: [
    {
      name: 'kill darling',
      fn: killDarling
    },
    {
      name: 'revive darling',
      fn: reviveDarling
    },
    {
      name: 'show darlings',
      fn: function () {
        neo.CurrentStory.setPref('darlings.hide', false);
        document.documentElement.classList.add('show_darlings');
      }
    },
    {
      name: 'hide darlings',
      fn: function () {
        neo.CurrentStory.setPref('darlings.hide', true);
        document.documentElement.classList.remove('show_darlings');
      }
    }
  ]
}

function killDarling() {
  var cursor = document.querySelector('nui-cursor');

  if (!cursor) {
    return;
  }

  var range = cursor.selectionRange;

  if (range && !range.collapsed) {
    log.debug('killing darling');
    var de = document.createElement('nui-darling');
    range.surroundContents(de);
    var cursor = de.cursor;
    cursor.parentNode.insertBefore(cursor, de.followingSibling);
  }
}

function reviveDarling() {
  var cursor = document.querySelector('nui-cursor');

  if (!cursor) {
    return;
  }

  var range = cursor.selectionRange;
  if (!range || range.collapsed) {
    if (cursor.previousSibling && cursor.previousSibling.tagName == 'NUI-DARLING') {
      cursor.previousSibling.unwrap();
    }

    if (cursor.nextSibling && cursor.nextSibling.tagName == 'NUI-DARLING') {
      cursor.nextSibling.unwrap();
    }
  } else {
    var da = [];

    for (var subrange = range.next(); subrange; subrange = range.next(subrange)) {
      if (subrange.startContainer.tagName == "NUI-DARLING") {
        da.push(subrange.startContainer);
      }
    }

    da.forEach(el => el.unwrap());
  }
}

function initTab() {
  var moduleLink = $('<link id="darlings_view" rel="import" href="' + path.join(__dirname, 'index.html') + '"></link>');
  $('head').append(moduleLink);

  document.addEventListener('open_story', e => {
    if (neo.CurrentStory.getPref('darlings.hide')) {
      document.documentElement.classList.remove('show_darlings');
    } else {
      document.documentElement.classList.add('show_darlings');
    }
  });
};

initTab();
